test/modules/TwinTester/ReportedPropertiesValidator.cs

```diff
@@ -89,46 +90,76 @@ async Task<TwinCollection> ValidatePropertiesFromTwinAsync(Twin receivedTwin)
         {
             TwinCollection propertyUpdatesFromTwin = receivedTwin.Properties.Reported;
             Dictionary<string, DateTime> reportedPropertiesUpdated = await this.storage.GetAllReportedPropertiesUpdatedAsync();
-            TwinCollection propertiesToRemoveFromTwin = new TwinCollection();
+            TwinCollection propertiesToRemove = new TwinCollection();
+
             foreach (KeyValuePair<string, DateTime> reportedPropertyUpdate in reportedPropertiesUpdated)
             {
                 string status;
                 if (propertyUpdatesFromTwin.Contains(reportedPropertyUpdate.Key))
                 {
-                    try
+                    if (this.reportedPropertyKeysFailedToReceiveWithinThreshold.Contains(reportedPropertyUpdate.Key))
                     {
-                        await this.storage.RemoveReportedPropertyUpdateAsync(reportedPropertyUpdate.Key);
+                        this.reportedPropertyKeysFailedToReceiveWithinThreshold.Remove(reportedPropertyUpdate.Key);
+                        status = $"{(int)StatusCode.ReportedPropertyReceivedAfterThreshold}: Successfully validated reported property update (exceeded failure threshold)";
                     }
-                    catch (Exception e)
+                    else
                     {
-                        Logger.LogError(e, $"Failed to remove validated reported property id {reportedPropertyUpdate.Key} from storage.");
-                        continue;
+                        status = $"{(int)StatusCode.ValidationSuccess}: Successfully validated reported property update";
                     }
 
-                    status = $"{(int)StatusCode.ValidationSuccess}: Successfully validated reported property update";
-                    Logger.LogInformation(status + $" {reportedPropertyUpdate.Key}");
+                    await this.HandleReportStatusAsync(status);
+                    Logger.LogInformation($"{status} for reported property update {reportedPropertyUpdate.Key}");
+                    propertiesToRemove[reportedPropertyUpdate.Key] = null; // will later be serialized as a twin update
                 }
-                else if (this.ExceedFailureThreshold(this.twinState, reportedPropertyUpdate.Value))
+                else if (!this.reportedPropertyKeysFailedToReceiveWithinThreshold.Contains(reportedPropertyUpdate.Key) &&
+                        this.ExceedFailureThreshold(this.twinState, reportedPropertyUpdate.Key, reportedPropertyUpdate.Value))
```

Comment:
super nit: I try to always make my bool names ask a question (i.e. DoesExceedFailureThreshold)

Reply:
👌